lib/event-libs/sdevent/sdevent.c

```diff
@@ -0,0 +1,400 @@
+    }
+
+    return 0;
+
+bail:
+    lws_pt_unlock(pt);
+    lws_context_unlock(pt->context);
+
+    return -1;
+}
+
+static void io_sd(struct lws *wsi, int flags) {
+
+    struct lws_context_per_thread *pt = &wsi->a.context->pt[(int)wsi->tsi];
+
+    // only manipulate if there is an event source and if the pt is still alive
+    if (!pt_to_priv_sd(pt)->io_loop || !wsi_to_priv_sd(wsi)->source || pt->is_destroyed) {
+        return;
+    }
+
+    // assert that the requested flags do not contain anything unexpected
+    if (!((flags & (LWS_EV_START | LWS_EV_STOP)) &&
+          (flags & (LWS_EV_READ | LWS_EV_WRITE)))) {
+        lwsl_err("%s: assert: flags %d", __func__, flags);
+        assert(0);
+    }
+
+    // we are overdoing a bit here, so it resembles the structure in libuv.c
+    if (flags & LWS_EV_START) {
+        if (flags & LWS_EV_WRITE)
+            wsi_to_priv_sd(wsi)->events |= EPOLLOUT;
+
+        if (flags & LWS_EV_READ)
+            wsi_to_priv_sd(wsi)->events |= EPOLLIN;
+
+        sd_event_source_set_io_events(wsi_to_priv_sd(wsi)->source, wsi_to_priv_sd(wsi)->events);
+        sd_event_source_set_enabled(wsi_to_priv_sd(wsi)->source, SD_EVENT_ONESHOT);
+    } else {
+        if (flags & LWS_EV_WRITE)
+            wsi_to_priv_sd(wsi)->events &= ~EPOLLOUT;
+
+        if (flags & LWS_EV_READ)
+            wsi_to_priv_sd(wsi)->events &= ~EPOLLIN;
+
+        sd_event_source_set_io_events(wsi_to_priv_sd(wsi)->source, wsi_to_priv_sd(wsi)->events);
+
+        if (!(wsi_to_priv_sd(wsi)->events & (EPOLLIN | EPOLLOUT)))
+            sd_event_source_set_enabled(wsi_to_priv_sd(wsi)->source, SD_EVENT_ONESHOT);
+        else
+            sd_event_source_set_enabled(wsi_to_priv_sd(wsi)->source, SD_EVENT_OFF);
+    }
+}
+
+static int init_vhost_listen_wsi_sd(struct lws *wsi) {
+    if (!wsi)
+        return 0;
+
+    struct lws_context_per_thread *pt = &wsi->a.context->pt[(int)wsi->tsi];
+
+    void *userdata = wsi;
+
+    sd_event_add_io(
+            pt_to_priv_sd(pt)->io_loop,
+            &wsi_to_priv_sd(wsi)->source,
+            wsi->desc.sockfd,
+            wsi_to_priv_sd(wsi)->events,
+            sock_accept_handler,
+            userdata
+    );
+
+    io_sd(wsi, LWS_EV_START | LWS_EV_READ);
+
+    return 0;
+}
+
+static int init_pt_sd(struct lws_context *context, void *_loop, int tsi) {
+    struct lws_context_per_thread *pt = &context->pt[tsi];
+    struct lws_pt_eventlibs_sdevent *ptpriv = pt_to_priv_sd(pt);
+
+    struct sd_event *loop = (struct sd_event *)_loop;
+    int first = 1;  // we are the first that create and initialize the loop
+
+    ptpriv->pt = pt;
+
+    // make sure we have an event loop
+    if (!ptpriv->io_loop) {
+        if (!loop) {
+            if (sd_event_default(&loop) < 0) {
+                lwsl_err("cannot get/create event loop, possibly out-of-memory\n");
+                return -1;
+            }
+            pt->event_loop_foreign = 0;
+        } else {
+            pt->event_loop_foreign = 1;
```

Comment:
We need to increase the ref count here.

```suggestion
            pt->event_loop_foreign = 1;
            sd_event_ref(loop);
```